tests/lib/hardware/libvirt.py

```diff
@@ -0,0 +1,300 @@
+<domain type='kvm'>
+<name>%(domain_name)s</name>
+<memory unit='KiB'>%(memory)s</memory>
+<currentMemory unit='KiB'>%(memory)s</currentMemory>
+<vcpu placement='static'>2</vcpu>
+<cpu mode='host-passthrough'>
+</cpu>
+<!--<cpu mode='host-model'>
+<feature policy='require' name='vmx'/>
+</cpu>-->
+<os>
+<type arch='x86_64' machine='pc-i440fx-2.1'>hvm</type>
+<boot dev='hd'/>
+</os>
+<on_poweroff>destroy</on_poweroff>
+<on_reboot>restart</on_reboot>
+<on_crash>restart</on_crash>
+<devices>
+<emulator>/usr/bin/qemu-system-x86_64</emulator>
+<disk type='file' device='disk'>
+<driver name='qemu' type='qcow2' cache='none'/>
+<source file='%(image)s'/>
+<target dev='vda' bus='virtio'/>
+</disk>
+<disk type='file' device='cdrom'>
+<driver name='qemu' type='raw' />
+<source file='%(cloud_init_seed)s'/>
+<target dev='sda' bus='sata'/>
+<readonly/>
+</disk>
+<controller type='virtio-serial' index='0'>
+<address type='pci' domain='0x0000' bus='0x00' slot='0x05' function='0x0'/>
+</controller>
+<interface type='network'>
+<source network='%(network_name)s'/>
+<model type='virtio'/>
+<address type='pci' domain='0x0000' bus='0x00' slot='0x03' function='0x0'/>
+</interface>
+<serial type='pty'>
+<target port='0'/>
+</serial>
+<console type='pty'>
+<target type='serial' port='0'/>
+</console>
+<channel type='spicevmc'>
+<target type='virtio' name='com.redhat.spice.0'/>
+<address type='virtio-serial' controller='0' bus='0' port='1'/>
+</channel>
+<input type='mouse' bus='ps2'/>
+<input type='keyboard' bus='ps2'/>
+<graphics type='spice' autoport='yes'/>
+<video>
+<model type='vga'/>
+</video>
+<redirdev bus='usb' type='spicevmc'>
+</redirdev>
+<memballoon model='virtio'>
+<address type='pci' domain='0x0000' bus='0x00' slot='0x06' function='0x0'/>
+</memballoon>
+</devices>
+</domain>
+        """ % {
+            "domain_name": domain_name, "image": image,
+            "cloud_init_seed": cloud_init_seed,
+            "network_name": network_name,
+            "memory": memory
+        }
+
+
+class Hardware(HardwareBase):
+    def __init__(self):
+        super().__init__()
+        self._network = self.conn.networkLookupByName(
+            config.PROVIDER_LIBVIRT_NETWORK)
+        if not self._network:
+            raise Exception(f'Can not get libvirt network '
+                            '{config.PROVIDER_LIBVIRT_NETWORK}')
+        logger.info(f"Got libvirt network {self._network.name()}")
+
+    def get_connection(self):
+        conn = libvirt.open(config.PROVIDER_LIBVIRT_CONNECTION)
+        if not conn:
+            raise Exception(f'Can not open libvirt connection '
+                            '{config.PROVIDER_LIBVIRT_CONNECTION}')
+        logger.debug(f"Got connection to libvirt: {conn}")
+        return conn
+
+    def _boot_node(self, name: str, role: NodeRole, tags: List[str]):
+        # get a fresh connection to avoid threading problems
+        conn = self.get_connection()
+        node = Node(name, role, tags, conn, self._network, 0, 8,
```

Comment:
(nit) we should document the VM sizes and/or make them configurable. (We make the OpenStack ones configurable through flavours).

Reply:
Done in an extra commit